File: behavioral-patterns/observer-pattern/src/test/java/SubscriberTests.java
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class SubscriberTests {

    private final ByteArrayOutputStream capture = new ByteArrayOutputStream();

    @BeforeEach
    public void setUp() {
        System.setOut(new PrintStream(capture));
    }

    @Test
    public void test1() {
        Publisher publisher = new Publisher(2);
        Observer observer1 = new Subscriber(1);
        Observer observer2 = new Subscriber(2);

        publisher.registerObserver(observer1);
        publisher.registerObserver(observer2);

        publisher.setState("Hello World!");
        Assertions.assertEquals(
                "Subscriber(1):\tSubscribe a message from publisher: Hello World!\n" +
                        "Subscriber(2):\tSubscribe a message from publisher: Hello World!",
                capture.toString().trim()
        );
        publisher.unregisterObserver(observer2);
        publisher.setState("Doing good?");
        Assertions.assertEquals(
                "Subscriber(1):\tSubscribe a message from publisher: Hello World!\n" +
                        "Subscriber(2):\tSubscribe a message from publisher: Hello World!\n" +
                        "Subscriber(1):\tSubscribe a message from publisher: Doing good?",  //  new message added...
                capture.toString().trim()
        );
    }
}

File: behavioral-patterns/observer-pattern/src/test/kotlin/SubscriberTests.kt
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.ByteArrayOutputStream
import java.io.PrintStream

class SubscriberTests {

    private val capture = ByteArrayOutputStream()

    @BeforeEach
    fun setup() {
        System.setOut(PrintStream(capture))
    }

    @Test
    fun test1() {
        val observer1 = TextViewSubscriber(1)
        val observer2 = TextViewSubscriber(2)
        val observer3 = TextViewSubscriber(3)

        val textView = TextView().apply {
            subscribers.add(observer1)
            subscribers.add(observer2)
            subscribers.add(observer3)
        }

        with(textView) {
            text = "Hello world!"
            Assertions.assertEquals(
                "Subscriber(1):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(2):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(3):\tText is changed <empty> -> Hello world!\n",
                capture.toString()
            )

            subscribers.remove(observer1)
            text = "Doing good?"
            Assertions.assertEquals(
                "Subscriber(1):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(2):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(3):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(2):\tText is changed Hello world! -> Doing good?\n" +
                        "Subscriber(3):\tText is changed Hello world! -> Doing good?\n",
                capture.toString()
            )

            subscribers.remove(observer2)
            text = "Goodbye friends!"
            Assertions.assertEquals(
                "Subscriber(1):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(2):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(3):\tText is changed <empty> -> Hello world!\n" +
                        "Subscriber(2):\tText is changed Hello world! -> Doing good?\n" +
                        "Subscriber(3):\tText is changed Hello world! -> Doing good?\n" +
                        "Subscriber(3):\tText is changed Doing good? -> Goodbye friends!\n",
                capture.toString()
            )
        }

    }
}